src/tribol/search/InterfacePairFinder.cpp

```diff
@@ -1,643 +1,873 @@
-
-    integer meshId2 = cs->getMeshId2();
-    MeshData const & meshData2 = meshManager.GetMeshInstance(meshId2);
-    integer mesh2NumElems = meshData2.m_numCells;
-
-    int numPairs = mesh1NumElems * mesh2NumElems;
-
-    InterfacePairs* contactPairs = cs->getInterfacePairs();
-    contactPairs->clear();
-    contactPairs->reserve( numPairs );
+template<int D, class ExecSpace>
+class BvhSearch : public SearchBase
+{
+public:
+  using BVHT = axom::spin::BVH<D, ExecSpace, RealT>;
+  using BoxT = typename BVHT::BoxType;
+  using PointT = primal::Point<RealT, D>;
+  using RayT = primal::Ray<RealT, D>;
+  using VectorT = primal::Vector<RealT, D>;
+  using AtomicPolicy = typename axom::execution_space<ExecSpace>::atomic_policy;
+
+  /*!
+  * Constructs a BvhSearch instance over CouplingScheme \a couplingScheme
+  * \pre couplingScheme is not null
+  */
+  BvhSearch(CouplingScheme* coupling_scheme)
+    : m_coupling_scheme( coupling_scheme )
+    , m_mesh1( m_coupling_scheme->getMesh1() )
+    , m_mesh2( m_coupling_scheme->getMesh2() )
+    , m_boxes1( axom::ArrayOptions::Uninitialized{},
+                m_mesh1.numberOfElements(),
+                m_mesh1.numberOfElements(),
+                m_coupling_scheme->getAllocatorId() )
+    , m_boxes2( axom::ArrayOptions::Uninitialized{},
+                m_mesh2.numberOfElements(),
+                m_mesh2.numberOfElements(),
+                m_coupling_scheme->getAllocatorId() )
+    , m_candidates( axom::ArrayOptions::Uninitialized{},
+                    0,
+                    0,
+                    m_coupling_scheme->getAllocatorId() )
+    , m_offsets( axom::ArrayOptions::Uninitialized{},
+                 m_mesh2.numberOfElements(),
+                 m_mesh2.numberOfElements(),
+                 m_coupling_scheme->getAllocatorId() )
+    , m_counts( axom::ArrayOptions::Uninitialized{},
+                m_mesh2.numberOfElements(),
+                m_mesh2.numberOfElements(),
+                m_coupling_scheme->getAllocatorId() )
+  {}
+
+  /*!
+  * Allocate and fill bounding box arrays for each of the two meshes
+  */
+  void initialize() override
+  {
+    buildMeshBBoxes(m_boxes1, m_coupling_scheme->getMesh1());
+    buildMeshBBoxes(m_boxes2, m_coupling_scheme->getMesh2());
+  } // end initialize()
+   
+
+  /*!
+  * Use the BVH to find candidates in first mesh for each
+  * element in second mesh of coupling scheme.
+  */
+  void findInterfacePairs() override
+  {
+    // Build the BVH
+    BVHT bvh;
+    bvh.setAllocatorID(m_coupling_scheme->getAllocatorId());
+    bvh.initialize(m_boxes1.view(), m_boxes1.size());
+
+    // Search for intersecting bounding boxes
+    bvh.findBoundingBoxes(m_offsets.view(),
+                          m_counts.view(),
+                          m_candidates,
+                          m_mesh2.numberOfElements(),
+                          m_boxes2.view());
+
+    // Apply geom filter to check if intersecting bounding boxes are in contact
```

Comment:
Consider: Apply geom filter to check if pairs in intersecting bounding boxes are proximate.